dashboard/spec/javascripts/client_state_test.js

```diff
@@ -1,63 +1,79 @@
+/* global dashboard */
+
 //= require client_state
 //= require jquery.cookie
 
-describe("clientState#levelProgress", function() {
+describe("clientState#trackProgress", function() {
   beforeEach(function() {
     dashboard.clientState.reset();
   });
 
-  it("records level progress", function() {
+  it("records level progress and line counts when level is completed", function() {
     var state = dashboard.clientState;
-    state.levelProgress(100).should.equal(0);
-    state.setLevelProgress(100, 50);
-    state.levelProgress(100).should.equal(50);
-    state.levelProgress(200).should.equal(0);
-    state.allLevelsProgress().should.eql({"100": 50})
-    state.setLevelProgress(100, 100);
-    state.setLevelProgress(200, 50);
-    state.levelProgress(100).should.equal(100);
-    state.levelProgress(200).should.equal(50);
-    state.allLevelsProgress().should.eql({"100": 100, "200": 50})
-  });
+    state.levelProgress(1).should.equal(0);
+    state.levelProgress(2).should.equal(0);
+    state.lines().should.equal(0);
 
-  it("handles malformed level progress cookies", function() {
-    $.cookie('progress', '',  {expires: 365})
-    var state = dashboard.clientState;
-    state.levelProgress(100).should.equal(0);
+    //User has passed a level with optimal solution
+    state.trackProgress(true, 5, 100, 1);
+    state.levelProgress(1).should.equal(100);
+    state.levelProgress(2).should.equal(0);
+    state.lines().should.equal(5);
 
-    $.cookie('progress', '[(*', {expires: 365})
```

Comment:
Could we preserve the malformed cookie test?  Good to know we handle this gracefully.